File: Data Structures & Algorithms/time-based-key-value-store/submission-4.py
class TimeMap:

    def __init__(self):
        self.key_value = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.key_value:
            self.key_value[key] = []
        self.key_value[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.key_value:
            return ""
        values = self.key_value[key]
        res = ""
        l, r = 0, len(values) - 1
        while l <= r:
            mid = (l + r) // 2
            if values[mid][0] <= timestamp:
                res = values[mid][1]
                l = mid + 1
            else:
                r = mid - 1
        return res

File: Data Structures & Algorithms/sqrtx/submission-4.java
class Solution {
    public int mySqrt(int x) {
        int l = 0;
        int r = x;
        int res = 0;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            long sqr = (long) mid * mid;
            if (sqr == x) {
                return mid;
            } else if (sqr < x) {
                l = mid + 1;
                res = mid;
            } else {
                r = mid - 1;
            }
        }
        return res;
    }
}